### PR TITLE
fix(noise-curvature): skip zero ema_decay 0*inf on curvature EMAs

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
## Summary
- Add `_skip_zero_scale` so `ema_decay=0` does not multiply poisoned `curvature_mean` / `curvature_variance` into NaN during controller updates.
- Regression test: poisoned infinite EMA history with `ema_decay=0` still applies a finite controller step without diagnostic failure.

Made with Cursor. No Slop measured-run receipt (not an approved Codex or Claude Code client).

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Review by kenjohnscreates — APPROVED** [review]
## Disposition: APPROVE — clean on all six checks, with one framing correction that makes the case *stronger*

Independently reproduced on **macOS arm64 / CPython 3.13.5**, at exact head `fc089b8b7b2bb8954a8a6a894192205bad340f1a` against base `c9aba7b54dedd647f8bd5f5c7bf6780b1413b676`, in isolated worktrees with `PYTHONPATH` pinned per tree (module resolution verified on both sides, so neither tree could import the other's copy).

I went looking for the failure modes that have bitten this series elsewhere — a vacuous regression test, a numeric drift on the untouched branch, a sibling site left behind. None of them are here.

### The six checks

1. **The regression test is genuine.** Copied onto pristine `main` and run against the unfixed library, it fails:
   ```
   FAILED tests/test_noise_curvature_ipmnist.py::test_zero_ema_decay_does_not_multiply_inf_curvature_stats
   E  assert 0 == 1  ... controller_count=Array(0), diagnostic_failures=Array(1)
   1 failed, 9 deselected
   ```
2. **Paired full-file runs**, `-q -p no:randomly`: `main` **9 passed** → this head **10 passed**. Zero regressions; the delta is exactly the added test.
3. **Bit-exactness on the branch the fix does not change.** Four `noise_curvature_step` iterations at three decays, hashing raw bytes of `curvature_mean`, `curvature_variance` and every parameter leaf:
   ```
   ema_decay=0.99  bbc64de17c796325     ema_decay=0.9  b8533d64c505382e     ema_decay=0.999  fc968446931c6f3c
   ```
   byte-identical on both sides. This is the property I care about most in this repo given the sealed `protocol_sha256` fingerprints (cf. #1987), and it holds. Helped by the fact that `decay` was *already* `jnp.asarray(config.ema_decay, dtype=jnp.float32)` on `main`, so wrapping it changes no evaluation domain — worth contrasting with #2554 in this same series, where introducing an array alias moved `(1 - beta)` from float64 into float32 and shifted the shipped default by 2281 ULP.
4. **Completeness — the file is closed.** The only sibling of this shape at line 312 is `beta * gradient_norm * volatility`, which is not an EMA over prior state. Lines 457 and 461 were the two sites and both are guarded.
   The asymmetry of guarding `decay * state.X` but leaving `(1.0 - decay) * y` unguarded is **correct, not an oversight**: `__post_init__` rejects `ema_decay == 1.0` explicitly (lines 140–142, `"ema_decay must lie in [0,1)"`), so the mirror-image `0 * inf` is unconstructible. I checked this rather than assuming it.
5. **Reachability — genuine, not latent.** `ema_decay` defaults to `0.9` but is validated only as `[0, 1)`, so `NoiseCurvatureConfig(ema_decay=0.0)` is a legal configuration a caller can construct. That puts this ahead of most of the series.
6. **Cross-PR overlap.** The only other open PR matching "noise_curvature" is #2300, which touches `alberta_framework/evaluation/noise_curvature_campaign.py`, not this module. The completeness claim is unconfounded.

`ruff check .` clean. `mypy` reports only the pre-existing macOS-only `O_TMPFILE` error that is equally present on pristine `main` — nothing attributable to this diff.

### Framing correction: `main` does not currently corrupt anything — it *refuses*

The PR body reads as though poisoned curvature statistics propagate today. They do not, and the reason is visible in the failure output above: `diagnostic_failures=1`, `controller_count=0`.

`main` already has a fail-closed finiteness guard at lines 498–500:

```python
valid = ... & jnp.all(jnp.isfinite(mean_uncorrected))
            & jnp.all(jnp.isfinite(variance_uncorrected))
            & jnp.all(jnp.isfinite(learning_rates))
...
diagnostic_failures=state.diagnostic_failures + (~valid).astype(jnp.int32)
```

so the `NaN` from `0 * inf` is detected and the controller update is discarded. Nothing poisoned reaches published state.

That makes the real defect **different from, and better than, the one claimed**: with `ema_decay=0.0` — a legal configuration — the controller is *permanently* disabled, silently, because it trips its own diagnostic on every step and charges a `diagnostic_failures` count for it. After this patch the same run gives `controller_count=1, diagnostic_failures=0` and the controller actually runs, with the mathematically correct `decay=0` limit (`mean = normalized_sharpness`).

I would rewrite the summary along those lines. "Prevents a NaN from poisoning curvature stats" is defensible but overstates the hazard and understates the fix; "makes `ema_decay=0` a usable configuration instead of one that silently disables the controller via a spurious diagnostic failure" is what the code actually does, and is a cleaner thing for a maintainer to accept.

### Reproduction

```bash
git fetch origin pull/2551/head:pr2551
git worktree add /tmp/wt-main c9aba7b5 --detach
git worktree add /tmp/wt-2551 pr2551 --detach
cp /tmp/wt-2551/tests/test_noise_curvature_ipmnist.py /tmp/wt-main/tests/
cd /tmp/wt-main && PYTHONPATH=$PWD python -m pytest tests/test_noise_curvature_ipmnist.py -q -p no:randomly \
  -k zero_ema_decay_does_not_multiply_inf_curvature
```

Static review plus local execution in throwaway worktrees; I did not run contributor code outside them.

**Comment by lalalune** [comment]
Closing: the curvature controller validates mean_uncorrected and variance_uncorrected before adopting either, preserving its prior finite statistics when a diagnostic overflows. The added test manually installs infinities that valid controller transitions cannot retain. No reachable finite-state failure or development benefit is established.
